### ISSUE TITLE
`type: 'number'` doesn't seem to actually cast properly

### ISSUE THREAD

**Issue by shadowRR** [jagi/meteor-astronomy#91]
Was wondering if i'm doing anything wrong with the type selection on a schema, since i'm using `type: "number"`, which reading from the documentation, should cast the value as integer (or float), but it's still saved as string in mongo ? Do i have to do something else to get it to be saved as integer or float ?

**Comment by awei01** [comment]
Do you have a code sample?

**Comment by awei01** [comment]
```
var Model = Astro.createClass({ 
    name: "model",
    fields: { 
        someNumber: "number"
    }
});
var instance = new Model({ someNumber: "3" });
```
should work.

**Comment by shadowRR** [comment]
Actually, it works when not in a nested object, but when used like this:
```
'projected.$.amount': {
            type: 'number',
            validators: [
                Validators.required(),
                Validators.number()
            ]
        }
```
It does not work.

**Comment by jagi** [comment]
If it didn't work, it will work in the new version :). It's because of the entirely new model for nested fields which makes code clean and prevents from making bugs. In the new version, your schema would look like follows:

```js
fields: {
  projected: {
    type: 'array',
    nestedType: 'number',
    validators: Validators.each(
      Validators.number()
    )
  }
}
```

It's pointless for me to fix it right now in the current version, so forgive me but you have to wait for 1.0 pre release :)

**Comment by shadowRR** [comment]
Ok, but in my case (if a understand you're example properly), it's not an array of number, but an array of objects for which one of its field is a `number`. So `nestedType` wouldn't work.

Anyway, it's not preventing my app from working properly, it was just a weird behavior ;)

**Comment by jagi** [comment]
Oh I see, sorry. So it would look like follows:

```js
Projected = Astro.Class({
  name: 'Projected',
  fields: {
    amount: {
      type: 'number',
      required: true,
      validators: Validators.number()
    }
  }
});

/* ... */

fields: {
  projected: {
    type: 'array',
    class: 'Projected'
  }
}
```

As you can see I've created "nested" class. You can store many instances of the `Projected` class in the `projected` field which is array.

In this release, in the `array` type you can specify nested type or nested class.

**Comment by shadowRR** [comment]
Ok, sound quite clean. But i'll have to actually create class outside of the main one from now on ? I won't be able to wrap it in one bloc like it's the case right now ?

**Comment by jagi** [comment]
I'm thinking about such solution. It would be easy to do:

```js
fields: {
  projected: {
    type: 'array',
    class: {
      name: 'Projected',
      fields: {
        amount: {
          type: 'number',
          required: true,
          validators: Validators.number()
        }
      }
    }
  }
}
```

This would create a class on the fly. It's more concise (both solution would work). What do you think?

I was also thinking about renaming `nestedType` and `class` properties and create only one property `nested` or `nest`:

```js
fieldName: { // Array of strings
  type: 'array',
  nested: 'string'
}

// or

fieldName: { // Array of instances of the ClassName class
  type: 'array',
  nested: 'ClassName'
}

// or anonymous class

fieldName: { // Array of instances of the anonymous class
  type: 'array',
  nested: {
    name: 'ClassName',
    fields: {
      /* list of fields */
    }
  }
}
```

It would work as follows.
- Check value of the `nested` property
- `if (_.isString(nested))`
  - then look for the type named as the value of the `nested` property
  - If type has not been found then look for the class named as the value of the `nested` property
- `else if (_.isObject(nested))`
  - try creating class on the fly
- else
  - rise error

**Comment by shadowRR** [comment]
On your first example, you're declaring you're anonymous class with the `class` keywords, and with `nested` on the second one ? I would find the `class` confusing...But i'm guessing that was just a typo ?

I definitely like you're `nested` idea. Would make a clean schema, no more repeating yourself with `array.$.nestedProp`. Everything from the same property.

**Comment by cwohlman** [comment]
+1 for nested classes

**Comment by jagi** [comment]
@shadowRR I've just presented different ways of declaring it. So, you think that naming it `nested` is the most intuitive?

@cwohlman it's already implemented in new version :). I just want to know your opinions on how to declare it :)

**Comment by shadowRR** [comment]
@jagi yep, `nested` is better IMHO, as `class` is just confusing when declared inside of a schema.

**Comment by jagi** [comment]
Ok, maybe you're right that it's confusing. Having only one property will be easier to learn:
You want to nest some type then use the `nest` property for that purpose.

> I definitely like you're nested idea. Would make a clean schema, no more repeating yourself with array.$.nestedProp. Everything from the same property.

Yes, that was the main idea :). At the beginning I've borrowed a solution from Aldeed for declaring nested validators. However, after playing with it for longer time and seeing how many problems it causes, I've decided to figure out something better. And I'm very happy with how it works.

**Comment by cwohlman** [comment]
A couple of other suggestions for the name/declaration: `subDoc` `itemType` `schema` `subschema` `itemClass`.

Although it's more typing I'm in favor of the `itemType` property because it's more semantic.

```
fieldName: { // Array of instances of the anonymous class
  type: 'array',
  itemType: {
    name: 'ClassName',
    fields: {
      /* list of fields */
    }
  }
}
fieldName: { // Object instance of the anonymous class
  type: 'object',
  itemType: {
    name: 'ClassName',
    fields: {
      /* list of fields */
    }
  }
}
```

**Comment by shadowRR** [comment]
I don't actually agree with you on this @cwohlman. `itemType` feels generic but rather unclear which isn't the case with `nested`.

**Comment by cwohlman** [comment]
The way I look at it `type` is the type of the field, `itemType` is the
type of the items in the field, e.g. the object or objects which will be in
the field.

The reason I don't like `nested` is because it feels like a boolean field
to me.

**Comment by jagi** [comment]
So maybe `nestedType`?

**Comment by shadowRR** [comment]
I'll get used to whatever you choose anyway, but i still prefer `nested`.

@cwohlman there's no concept of item(s) in astronomy, but i do get your point on boolean field. `nestedType` would be ok with you ?

**Comment by jagi** [comment]
Take in mind that it can not only be nested array but also nested object. So we can store one or more elements. For me `nested` or `nestedType` seems to be the best description of what it's for. We nest some value in another.

I agree that `nested` may sound like the name of a boolean variable. However for such cases it should be `isNested` - to be precise. We could also name it `nest` - imperative form. We tell it to nest this value. But I'm not fan of it.

Maybe discussion about naming property seems trivial but I think many people care about it :-). And it's easier to learn the tool if names are self explanatory.

**Comment by cwohlman** [comment]
> Maybe discussion about naming property seems trivial but I think many people care about it :-). And it's easier to learn the tool if names are self explanatory.

:+1:  I agree that it's important what we name the property, so that the api make intuitive sense.

> Take in mind that it can not only be nested array but also nested object. So we can store one or more elements. For me nested or nestedType seems to be the best description of what it's for. We nest some value in another.

:+1: I'm for using `nestedType`

I'm wondering if we're going to call the field `nestedType` if we should allow a `Class` to be the value for the field `type`, which would basically make `type: MyClass` shorthand for `type: 'object', nestedType: MyClass`.

**Comment by jagi** [comment]
> I'm wondering if we're going to call the field nestedType if we should allow a Class to be the value for the field type, which would basically make type: MyClass shorthand for type: 'object', nestedType: MyClass.

It used to work that way at the beginning when I started working on this feature but for some reason I had to abandon it. It would take to much time to explain but I came to this conclusion after several weeks of testing. They way it works right now is correct :)

I've decided to name it `nested` - sorry :). But it's just for a pre release version. If developers will say that it's not intuitive or it needs a change, I will take it into consideration :). It's just a name that is easy to change in final version :).
